tools/eslint-plugin-azure-sdk/src/rules/ts-apiextractor-json-types.ts

```diff
@@ -0,0 +1,72 @@
+/**
+ * @file Rule to force the inclusion of type declarations in the package.
+ * @author Arpan Laha
+ * @author Will Temple
+ */
+
+import { Rule } from "eslint";
+import { Literal, Property } from "estree";
+import { getRuleMetaData, getVerifiers, stripPath } from "../utils";
+import { stripFileName } from "../utils/verifiers";
+
+//------------------------------------------------------------------------------
+// Rule Definition
+//------------------------------------------------------------------------------
+
+export = {
+  meta: getRuleMetaData(
+    "ts-apiextractor-json-types",
+    "force api-extractor.json to configure types in a consistent way"
+  ),
+  create: (context: Rule.RuleContext): Rule.RuleListener => {
+    const verifiers = getVerifiers(context, {
+      outer: "dtsRollup",
+      inner: "publicTrimmedFilePath",
+      expected: false
+    });
+    const fileName = context.getFilename();
+    return stripPath(fileName) === "api-extractor.json"
+      ? ({
+          // callback functions
+          "ExpressionStatement > ObjectExpression": verifiers.existsInFile,
+          // check to see if dtsRollup.publicTrimmedFilePath is defined
+          "ExpressionStatement > ObjectExpression > Property[key.value='dtsRollup']":
+            verifiers.isMemberOf,
+
+          // check the node corresponding to types to see if its value is a TypeScript declaration file
+          "ExpressionStatement > ObjectExpression > Property[key.value='dtsRollup'] > ObjectExpression > Property[key.value='publicTrimmedFilePath']": (
+            node: Property
+          ): void => {
+            if (
+              node.value.type !== "Literal" ||
+              typeof node.value.value !== "string"
+            ) {
+              context.report({
+                node: node.value,
+                message: ".d.ts rollup path is not set to a string"
+              });
+            }
+
+            const nodeValue = node.value as Literal;
```

Comment:
Can we avoid doing all this type casting with `as` somehow? When you use `as` to narrow a type, you're not actually checking it to make sure it's the specific type. 

Prefer using user defined type guards if you need to narrow deterministically: https://www.typescriptlang.org/docs/handbook/advanced-types.html#user-defined-type-guards

Reply:
Yes. I can change this in the other rule that I modified as well (this rule is almost a direct copy of `ts-package-json-types`)

Looking at it a little more closely, I think there's a bug in both rules that the rule should return in the `if (node.value.type !== "Literal") ...` since it doesn't make sense to cast it to Literal afterwords. If that block returns or if we wrap the rest of the body in an `else`,, TypeScript can refine the type to `SimpleLiteral` with `value : string` for us and no cast or manual guard is necessary in the rest of the body anyway.

I will add tests for that case and make some changes to both rules as a fixup.